src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -28,8 +29,37 @@
                       SerializingFunction)
 from ._utils import _timeout_to_deadline
 
+_TIMEOUT_WAIT_FOR_CLOSE_ONGOING_CALLS_SEC = 0.1
 _IMMUTABLE_EMPTY_TUPLE = tuple()
 
+_LOGGER = logging.getLogger(__name__)
+
+
+class _OngoingCalls:
+    """Internal class used for have visibility of the ongoing calls."""
+
+    _calls: Sequence[_base_call.RpcContext]
+
+    def __init__(self):
+        self._calls = []
+
+    def _remove_call(self, call: _base_call.RpcContext):
+        self._calls.remove(call)
```

Comment:
`list` is implemented as a dynamic array. [`remove` is going to be `O(n)`](https://github.com/python/cpython/blob/master/Objects/listobject.c#L2619).  Seems like `set` would be more optimal here. [It's implemented as a dynamically resized hashmap](https://github.com/python/cpython/blob/master/Objects/setobject.c). Insert should still be amortized `O(1)` and remove will be `O(1)`.

Reply:
Yeps, I thought the same as you  ... the only reason because I've made this using a list was for maintaining the order that the calls were created, so if the user made to calls with the list we will guarantee that the cancellation will happen in some kind of order. But seeing that

1) Current implementation iterates from left to right, so a weird preserving order will be meet
2) When the grace period is finished, cancellation happens within the same loop iteration so calls won't have any chance to react on that.

Moving to use a set.

Thanks for the feedback.